pkg/instance/service_test.go

```diff
@@ -2,16 +2,206 @@ package instance
+		service := NewService(nil, nil, stackService, nil)
+		deployment := &model.Deployment{
+			Instances: []*model.DeploymentInstance{
+				{
+					StackName: "name",
+					Parameters: map[string]model.DeploymentInstanceParameter{
+						"parameter": {
+							ParameterName: "parameter",
+						},
+					},
+				},
+			},
+		}
+
+		err := service.resolveParameters(deployment)
+
+		require.ErrorContains(t, err, "parameter not found on stack: parameter")
+	})
+
+	t.Run("ResolveParameters", func(t *testing.T) {
+		defaultValue1 := "default value used"
+		defaultValue2 := "default value not user"
+		stackA := model.Stack{
+			Name: "stack-a",
+			Parameters: map[string]model.StackParameter{
+				"parameter-a": {
+					DefaultValue: &defaultValue1,
+				},
+				"parameter-b": {
+					DefaultValue: &defaultValue2,
+				},
+				"parameter-c": {},
+			},
+		}
+		stacks := stack.Stacks{
+			"stack-a": stackA,
+		}
+		stackService := stack.NewService(stacks)
+		service := NewService(nil, nil, stackService, nil)
+		deployment := &model.Deployment{
+			Instances: []*model.DeploymentInstance{
+				{
+					StackName: "stack-a",
+					Parameters: map[string]model.DeploymentInstanceParameter{
+						"parameter-b": {
+							ParameterName: "parameter-b",
+							Value:         "default value overwritten by user",
+						},
+						"parameter-c": {
+							ParameterName: "parameter-c",
+							Value:         "some value",
+						},
+					},
+				},
+			},
+		}
+
+		err := service.resolveParameters(deployment)
+
+		require.NoError(t, err)
+		want := []*model.DeploymentInstance{
+			{
+				StackName: "stack-a",
+				Parameters: map[string]model.DeploymentInstanceParameter{
+					"parameter-a": {
+						ParameterName: "parameter-a",
+						Value:         "default value used",
+					},
+					"parameter-b": {
+						ParameterName: "parameter-b",
+						Value:         "default value overwritten by user",
+					},
+					"parameter-c": {
+						ParameterName: "parameter-c",
+						Value:         "some value",
+					},
+				},
+			},
+		}
+		assert.ElementsMatch(t, want, deployment.Instances)
+	})
+
+	t.Run("ResolveParameterUsingProvider", func(t *testing.T) {
+		stackA := model.Stack{
+			Name: "stack-a",
+			ParameterProviders: model.ParameterProviders{
+				"provider-parameter": model.ParameterProviderFunc(func(instance model.DeploymentInstance) (string, error) {
+					return fmt.Sprintf("%s-%s", instance.Name, instance.GroupName), nil
+				}),
+			},
+		}
+		stackB := model.Stack{
+			Name: "stack-b",
+			Parameters: map[string]model.StackParameter{
+				"provider-parameter": {
+					Consumed: true,
+				},
+			},
+			Requires: []model.Stack{stackA},
+		}
+		stacks := stack.Stacks{
+			"stack-a": stackA,
+			"stack-b": stackB,
+		}
+		stackService := stack.NewService(stacks)
+		service := NewService(nil, nil, stackService, nil)
+		deployment := &model.Deployment{
+			Instances: []*model.DeploymentInstance{
+				{
+					Name:       "name",
```

Comment:
you could make them `name-a`, `group-a` just to make it more clear in the expected parameter that the values come from an instance of stack a.